zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
+    // if combination of shift[1..32] == 0
+    // shift_overflow will be equal to 0, otherwise 1.
+    shift_overflow: Cell<F>,
+    // is_zero will check combination of shift[1..32] == 0
+    is_zero: IsZeroGadget<F>,
+}
+impl<F: FieldExt> ShrWordsGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        a: util::Word<F>,
+        shift: util::Word<F>,
+    ) -> Self {
+        let b = cb.query_word();
+        let a_slice_front = array_init::array_init(|_| cb.query_byte());
+        let a_slice_back = array_init::array_init(|_| cb.query_byte());
+        let shift_div64 = cb.query_cell();
+        let shift_mod64_div8 = cb.query_cell();
+        let shift_mod64_decpow = cb.query_cell();
+        let shift_mod64_pow = cb.query_cell();
+        let shift_mod8 = cb.query_cell();
+        let shift_overflow = cb.query_bool();
+
+        // check (combination of shift[1..32] == 0) == 1 - shift_overflow
+        let mut sum = 0.expr();
+        (1..32).for_each(|idx| sum = sum.clone() + shift.cells[idx].expr());
+        let is_zero = IsZeroGadget::construct(cb, sum);
+        cb.require_equal(
+            "shift_overflow == shift > 256 ",
+            shift_overflow.expr(),
+            1.expr() - is_zero.expr(),
+        );
+
+        // rename variable:
+        // shift_div64 :a
+        // shift_mod64_div8:b
+        // shift_mod8:c
+        // we split shift[0] to the equation:
+        // shift[0] == a * 64 + b * 8 + c
+        let shift_mod64 =
+            8.expr() * shift_mod64_div8.expr() + shift_mod8.expr();
+        cb.require_equal(
+            "shift[0] == shift_div64 * 64 + shift_mod64_div8 * 8 + shift_mod8",
+            shift.cells[0].expr(),
+            shift_div64.expr() * 64.expr() + shift_mod64.clone(),
+        );
+
+        // merge 8 8-bit cell for a 64-bit expression
+        // for a, a_slice_front, a_slice_back, b
+        let mut a_digits = vec![];
+        let mut a_slice_front_digits = vec![];
+        let mut a_slice_back_digits = vec![];
+        let mut b_digits = vec![];
+        for virtual_idx in 0..4 {
+            let now_idx = (virtual_idx * 8) as usize;
+            a_digits.push(from_bytes::expr(&a.cells[now_idx..now_idx + 8]));
+            a_slice_back_digits
+                .push(from_bytes::expr(&a_slice_back[now_idx..now_idx + 8]));
+            a_slice_front_digits
+                .push(from_bytes::expr(&a_slice_front[now_idx..now_idx + 8]));
+            b_digits.push(from_bytes::expr(&b.cells[now_idx..now_idx + 8]));
+        }
+
+        // check combination of a_slice_back_digits and a_slice_front_digits
+        // == b_digits
+        let mut shr_constraints =
+            (0..4).map(|_| 0.expr()).collect::<Vec<Expression<F>>>();
+        for transplacement in (0_usize)..(4_usize) {
+            // generate the polynomial depends on the shift_div64
+            let select_transplacement_polynomial =
+                generate_lagrange_base_polynomial(
```

Comment:
You can replace the lagrange poly by a lookup table as you only have 64 * 4 = 256 cases in total. So you can avoid the high degree expression.